#EXTRACCION DE DATOS,
bajamos la data del lugar de interes
lo guardamos y subimos a drive

#Leemos el documento y nos conectamos a la biblioteca pandas
#creamos el dataframe con una variable

In [7]:
import pandas as pd

In [16]:
df_telcom = pd.read_json('/content/drive/MyDrive/reto2_ecomerce_alura-oracle/TelecomX_Data.json')
df_telcom.head()
#Nos damos cuenta que  costumer, phone, internet, account, tienen información  anidadadatos_telcom =

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

#visualizamos los datos anidados

In [21]:
df_telcom.info()
df_telcom.head(4)
df_telcom.columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


Index(['customerID', 'Churn', 'customer', 'phone', 'internet', 'account'], dtype='object')

In [29]:
df_customer = pd.json_normalize(df_telcom['customer'])
df_phone = pd.json_normalize(df_telcom['phone'])
df_internet = pd.json_normalize(df_telcom['internet'])
df_account = pd.json_normalize(df_telcom['account'])


In [32]:
df_normalizados = pd.concat([df_customer, df_phone, df_internet, df_account], axis=1)
df_normalizados.head()
df_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7267 non-null   object 
 1   SeniorCitizen     7267 non-null   int64  
 2   Partner           7267 non-null   object 
 3   Dependents        7267 non-null   object 
 4   tenure            7267 non-null   int64  
 5   PhoneService      7267 non-null   object 
 6   MultipleLines     7267 non-null   object 
 7   InternetService   7267 non-null   object 
 8   OnlineSecurity    7267 non-null   object 
 9   OnlineBackup      7267 non-null   object 
 10  DeviceProtection  7267 non-null   object 
 11  TechSupport       7267 non-null   object 
 12  StreamingTV       7267 non-null   object 
 13  StreamingMovies   7267 non-null   object 
 14  Contract          7267 non-null   object 
 15  PaperlessBilling  7267 non-null   object 
 16  PaymentMethod     7267 non-null   object 


In [35]:
#convertimos a numerico
print(df_normalizados.columns.tolist())


['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Charges.Monthly', 'Charges.Total']


In [38]:
df_normalizados.columns = df_normalizados.columns.str.replace('.', '_', regex=False)
df_normalizados.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes        Yes       9          Yes   
1    Male              0      No         No       9          Yes   
2    Male              0      No         No       4          Yes   
3    Male              1     Yes         No      13          Yes   
4  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0            No             DSL             No          Yes               No   
1           Yes             DSL             No           No               No   
2            No     Fiber optic             No           No              Yes   
3            No     Fiber optic             No          Yes              Yes   
4            No     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   
3          No         Yes             Yes  Month-to-month              Yes   
4         Yes         Yes              No  Month-to-month              Yes   

      PaymentMethod  Charges_Monthly Charges_Total  
0      Mailed check             65.6         593.3  
1      Mailed check             59.9         542.4  
2  Electronic check             73.9        280.85  
3  Electronic check             98.0       1237.85  
4      Mailed check             83.9         267.4

In [40]:
df_normalizados['Charges_Total'] = pd.to_numeric(df_normalizados['Charges_Total'], errors='coerce')
df_normalizados.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes        Yes       9          Yes   
1    Male              0      No         No       9          Yes   
2    Male              0      No         No       4          Yes   
3    Male              1     Yes         No      13          Yes   
4  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0            No             DSL             No          Yes               No   
1           Yes             DSL             No           No               No   
2            No     Fiber optic             No           No              Yes   
3            No     Fiber optic             No          Yes              Yes   
4            No     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   
3          No         Yes             Yes  Month-to-month              Yes   
4         Yes         Yes              No  Month-to-month              Yes   

      PaymentMethod  Charges_Monthly  Charges_Total  
0      Mailed check             65.6         593.30  
1      Mailed check             59.9         542.40  
2  Electronic check             73.9         280.85  
3  Electronic check             98.0        1237.85  
4      Mailed check             83.9         267.40

In [55]:
df_normalizados.shape


(6992, 19)

##vemos si hay duplicados

In [46]:
df_normalizados.duplicated().sum()


np.int64(264)

In [50]:
df_normalizados = df_normalizados.drop_duplicates()



#Información de dataframe
sabemos que tenemos 7267 filas y 19 columnas ya desanidando los datos de las columnas : customer,	phone,	internet,	account, y las dos columnas que no tenian información anidad CustomerID y Churn, hay 12 valores nulos en la columna Charges_Total, que aproximadamente representa un 0.15% de los datos, luego checamos duplicados y encontramos que habia 264 por lo que se eliminaron y con ello se fueron los null

#Verificamos los null si hay  eliminamos null

In [57]:
df_normalizados.isnull().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges_Monthly     0
Charges_Total       0
dtype: int64

##no hay null pero es posible que lo hubiera cuando los duplicados aun estaban, dejare la forma de eliminar como ejemplo los null
en este caso digamos que los null estaban en esta columna por loque eliminamos y luego verificamos     df_normalizados = df_normalizados.dropna(subset=['Charges_Total'])  verificamos usando df_normalizados['Charges_Total'].isnull().sum()
df_normalizados.shape


In [58]:
print(df_normalizados['Charges_Total'].dtype)


float64


In [60]:
print(df_normalizados['Charges_Monthly'].dtype)

float64


In [59]:
print(df_normalizados['Partner'].value_counts())


Partner
No     3599
Yes    3393
Name: count, dtype: int64


In [61]:
print(df_normalizados['Contract'].unique())
#vamos a unificar los formatos de escritura de esta columna

['One year' 'Month-to-month' 'Two year']


In [62]:
df_normalizados['Contract'] = df_normalizados['Contract'].str.strip().str.lower()


In [63]:
df_normalizados['Contract'] = df_normalizados['Contract'].map({
    'one year': '1 year',
    'month-to-month': 'month-to-month',
    'two year': '2 year'
})


In [64]:
print(df_normalizados['Contract'].unique())
print(df_normalizados['Contract'].value_counts())


['1 year' 'month-to-month' '2 year']
Contract
month-to-month    3835
2 year            1685
1 year            1472
Name: count, dtype: int64


In [65]:
print(df_normalizados['InternetService'].unique())

['DSL' 'Fiber optic' 'No']


In [66]:
df_normalizados['InternetService'] = df_normalizados['InternetService'].str.strip().str.lower()


In [67]:
df_normalizados['InternetService'] = df_normalizados['InternetService'].map({
    'dsl': 'dsl',
    'fiber optic': 'fiber_optic',
    'no': 'no'
})


In [68]:
print(df_normalizados['InternetService'].unique())
print(df_normalizados['InternetService'].value_counts())


['dsl' 'fiber_optic' 'no']
InternetService
fiber_optic    3087
dsl            2413
no             1492
Name: count, dtype: int64


In [70]:
print(df_normalizados['PaymentMethod'].unique())


['mailed check' 'electronic check' 'credit card (automatic)'
 'bank transfer (automatic)']


In [69]:
df_normalizados['PaymentMethod'] = df_normalizados['PaymentMethod'].str.strip().str.lower()


In [71]:
# Limpieza básica (por si hubiera espacios)
df_normalizados['PaymentMethod'] = df_normalizados['PaymentMethod'].str.strip().str.lower()

# Mapear a nombres más cortos
df_normalizados['PaymentMethod'] = df_normalizados['PaymentMethod'].map({
    'mailed check': 'mailed_check',
    'electronic check': 'electronic_check',
    'credit card (automatic)': 'credit_card',
    'bank transfer (automatic)': 'bank_transfer'
})


In [72]:
print(df_normalizados['PaymentMethod'].unique())
print(df_normalizados['PaymentMethod'].value_counts())


['mailed_check' 'electronic_check' 'credit_card' 'bank_transfer']
PaymentMethod
electronic_check    2357
mailed_check        1572
bank_transfer       1542
credit_card         1521
Name: count, dtype: int64


#Estadistica descriptiva para detectar valores fuera de rango

In [87]:
print(df_normalizados.columns.tolist())


['Gender', 'Seniorcitizen', 'Partner', 'Dependents', 'Tenure', 'Phoneservice', 'Multiplelines', 'Internetservice', 'Onlinesecurity', 'Onlinebackup', 'Deviceprotection', 'Techsupport', 'Streamingtv', 'Streamingmovies', 'Contract', 'Paperlessbilling', 'Paymentmethod', 'Chargesmonthly', 'Chargestotal']


In [88]:
cols = df_normalizados.columns.tolist()

replacements = {
    'Seniorcitizen': 'Senior Citizen',
    'Phoneservice': 'Phone Service',
    'Multiplelines': 'Multiple Lines',
    'Internetservice': 'Internet Service',
    'Onlinesecurity': 'Online Security',
    'Onlinebackup': 'Online Backup',
    'Deviceprotection': 'Device Protection',
    'Techsupport': 'Tech Support',
    'Streamingtv': 'Streaming TV',
    'Streamingmovies': 'Streaming Movies',
    'Paperlessbilling': 'Paperless Billing',
    'Paymentmethod': 'Payment Method',
    'Chargesmonthly': 'Charges Monthly',
    'Chargestotal': 'Charges Total'
}

cols_corrected = [replacements.get(c, c) for c in cols]

df_normalizados.columns = cols_corrected

print(df_normalizados.columns.tolist())




['Gender', 'Senior Citizen', 'Partner', 'Dependents', 'Tenure', 'Phone Service', 'Multiple Lines', 'Internet Service', 'Online Security', 'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method', 'Charges Monthly', 'Charges Total']


#Guardaremos los taros impios

In [89]:
df_normalizados.to_csv('/content/TelecomX_Data_Limpio.csv', index=False)


In [90]:
#lo descargare a mi maquina
from google.colab import files
files.download('/content/TelecomX_Data_Limpio.csv')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# en esta parte cree el repositorio guit guithub